### PR TITLE
smooth scrolling

### PR DESCRIPTION
**Pull request by dtorop**
Allow smooth scrolling for the majority of the UI components. The rationale is twofold:

1. Smooth scrolling can make things nicer for trackpad users. It has been supported since GTK+ 3.4, and it seems worth taking advantage of these events.
2. GDK's Wayland backend will only generate smooth scroll events for trackpads. Therefore this is a step towards re-enabling the Wayland backend. See bug #11535.

Some things are still not smooth-scrolled:
1. GtkComboBox does not implement smooth scrolling, even though it is widely used in the UI. I'm not clear if this is a bug or a decision. The very-similar bauhaus comboboxes do now allow smooth scrolling to change their values. Does it make sense to disable the latter, for consistency? Or wrap the former to allow smooth scrolling?
2. The view's center container still gets up-or-not-up rather than a smooth scroll delta. Smooth scrolling actually looks great in the center container, but the events also have to be routed to masks and possibly iops. For now am keeping things integral and these changes contained.

I found and fixed a few minor bugs along the way, as noted in commit messages. I had to make some calls where the original implementation seemed off when clamping ranges. I changed the behavior of scrolling on control points of basecurve and tone curve (flipped up/down, added left/right), I hope for the better.

### CODE REVIEW

**Comment by dtorop** [comment]
One question is whether, with so much boilerplate to handle UP/DOWN vs SMOOTH scrolling, if it makes sense to have functions (say either `dt_ui_get_scroll_delta_y()` or `dt_iop_gui_get_scroll_delta_y()` and then another to get/accumulate integral scroll values) to cull out a bunch of this repetitive code.

**Comment by LebedevRI** [comment]
Not commenting about smooth scrolling at the moment.

> (say either dt_ui_get_scroll_delta_y() or dt_iop_gui_get_scroll_delta_y() and then another to get/accumulate integral scroll values) to cull out a bunch of this repetitive code.

Yeah, it does make sense. Duplication is not good.

**Comment by dtorop** [comment]
> Yeah, it does make sense. Duplication is not good.

I'll do that for this PR. It'll cut some lines of code...

**Comment by dtorop** [comment]
I coalesced the scroll handling into a couple functions, `dt_gui_get_scroll_unit_deltas` which returns whole-number deltas, `dt_gui_get_scroll_deltas` which returns fractional deltas. The former should produce the same scroll behavior as before this PR, except that trackpad scrolling will work under Wayland. The latter adds smooth scrolling for trackpads to both X11 and Wayland backends (and Quartz?). Behavior for scroll wheels should remain unchanged.

Also rebased to combine a couple earlier commits.

It would be possible to change all scroll handling to only call the unit-delta function, which would fix touchpad scrolling for Wayland but otherwise not change UI behavior. It would even be possible to be extremely conservative and via some conditionals/ifdefs only enable GDK_SMOOTH_SCROLL_MASK under Wayland, to leave other codepaths even more as-it-were. This is what the Scintilla developers did (https://sourceforge.net/p/scintilla/bugs/1901/).

File: src/iop/basecurve.c [hunk 1]
```diff
@@ -1792,22 +1792,15 @@ static gboolean _scrolled(GtkWidget *widget, GdkEventScroll *event, gpointer use
...
 
-  int handled = 0;
-  float dy = 0.0f;
-  if(event->direction == GDK_SCROLL_UP)
-  {
-    handled = 1;
-    dy = BASECURVE_DEFAULT_STEP;
-  }
-  if(event->direction == GDK_SCROLL_DOWN)
+  gdouble delta_x, delta_y;
```
Comment (LebedevRI) [inline src/iop/basecurve.c]: In basecurve module, when placing the mouse over a point on the curve, and using mousewheel to scroll, the scroll direction is inverted compared to the git master.

File: src/iop/tonecurve.c [hunk 2]
```diff
@@ -807,22 +807,15 @@ static gboolean _scrolled(GtkWidget *widget, GdkEventScroll *event, gpointer use
...
 
-  int handled = 0;
-  float dy = 0.0f;
-  if(event->direction == GDK_SCROLL_UP)
-  {
-    handled = 1;
-    dy = TONECURVE_DEFAULT_STEP;
-  }
-  if(event->direction == GDK_SCROLL_DOWN)
+  gdouble delta_x, delta_y;
```
Comment (LebedevRI) [inline src/iop/tonecurve.c]: In tonecurve module, when placing the mouse over a point on the curve, and using mousewheel to scroll, the scroll direction is inverted compared to the git master.

File: src/iop/basecurve.c (hunk 1 above)
Comment (dtorop) [inline src/iop/basecurve.c]: I added a commit to fix this in the mousewheel scrolling case, while trying when possible to keep the scrolling in the direction of any touch or trackpoint events. As comments note, the latter isn't always possible, so I'm erring on the side of keeping the mousewheel behavior consistent.

It feels right, when possible, for touch scrolling to move the point in the direction of the scrolll...

File: src/iop/tonecurve.c (hunk 2 above)
Comment (dtorop) [inline src/iop/tonecurve.c]: Fixed here as well... Neither of these convinces github I actually made the change requested, due to my having altered lines a bit later in the file...

File: src/iop/basecurve.c (hunk 1 above)
Comment (LebedevRI) [inline src/iop/basecurve.c]: > It feels right, when possible, for touch scrolling to move the point in the direction of the scrolll...

Yep, same is valid for mouse scroolwheel/etc

**Comment by dtorop** [comment]
> I will only be able to test mouse scrollwheel and keyboard arrows with X11.
> So if someone with touchpad/etc. wayland/etc can test too, it would be nice.

@LebedevRI: thank you for the thorough review! I'm in the opposite boat so far as testing. I don't have a good wheel mouse, but have been testing on Wayland and X11 with touchpad/trackpoint. My hope is to make the Wayland case not broken, with an added bonus of making touchpad interaction better. But clearly good not to break the X11/scrollwheel case, which must be what most people depend upon.

**Comment by dtorop** [comment]
Incidentally, to cross-reference, https://bugzilla.mozilla.org/show_bug.cgi?id=958868 is also a wealth of information about the fine points of handling `GDK_SCROLL_SMOOTH`.

File: src/iop/basecurve.c (hunk 1 above)
Comment (dtorop) [inline src/iop/basecurve.c]: >> It feels right, when possible, for touch scrolling to move the point in the direction of the scrolll...

> Yep, same is valid for mouse scroolwheel/etc

I don't have a scroll wheel mouse here to check what is right for it. I should make the scrolling for all cases as it is for touch/trackpad in latest commit, yes?

Comment (dtorop) [inline src/iop/basecurve.c]: I went ahead and added a commit to make all scroll events be for base/tonecurve be handled the same, whether mouse scrollwheel or touch etc. This makes the code simpler, but does change behavior from what is in git master. But this is a more sensible behavior for scrollwheel mice as well as the other cases?

Comment (houz) [inline src/iop/basecurve.c]: The current (inverted) behaviour is on purpose as it's what hanatos prefers. Please keep it that way in this PR. Re-evaluating that decision is something for another day.

Comment (dtorop) [inline src/iop/basecurve.c]: > The current (inverted) behaviour is on purpose as it's what hanatos prefers. Please keep it that way in this PR. Re-evaluating that decision is something for another day.

Thank you for clarification! I flipped it back.

One question is what to do about horizontal scrolls. Prior to this PR, they were just plain ignored. Now they are handled (presumably only from trackpad/trackpoint/touch) and their direction is inverted (to be consistent with vertical scrolling). Would it be better just to ignore them, such that scrolling would be used to fine-tune the y-value of points without messing up the x-axis? Alternately, some keyboard modifier (meta/alt?) could be used to switch from y-only scrolling to x and y scrolling?

**Comment by dtorop** [comment]
> Hm, this does not change anything here, unfortunately :(
> In these two iops, scrolling up still moves the point down...

@LebedevRI: I flipped it now! I was fuzzy until @houz wrote about what behavior is preferred. I amended the last commit (rewrote history) so that there wouldn't be a huge stack of commits fine-tuning this.

**Comment by dtorop** [comment]
One more change to base/tonecurve: I eliminated x-axis scrolling behavior. This means scrolling in these iops should be unchanged compared to git master, excepting that it will now be smooth on trackpads and work on Wayland.

I was using these modules today and realized that x-axis scrolling of points really adds nothing (though it feels sort of cool to scroll points side-to-side), and indeed loses something valuable: the ability to adjust y values of points without mucking up their x values.

**Comment by dtorop** [comment]
I've been using darktable with this PR applied for a bit more. One unintended consequence of smooth scrolling in tonecurve/basecurve/levels etc. is that it adds a lot of extra undo items when one makes a scroll adjustment. This requires lots of crtl-z-tapping to get back to a prior state. I'm wondering if this is enough of a reason to turn scrolling in these cases back to accumulating a delta of 1 before making an adjustment. Certainly for tools such as tonecurve/basecurve, one case use a ctrl-modifier to make fine adjustments if needed.

For equalizer-style tools, where scrolling alters the size of adjustment cursor, accumulating too many undo events isn't a worry.

**Comment by TurboGit** [comment]
I have a PR to smooth those undo.

See https://github.com/darktable-org/darktable/pull/1419

The idea is indeed to avoid recording too many intermediate state for a module. Please try it and tell me how it works on your side.

**Comment by dtorop** [comment]
> The idea is indeed to avoid recording too many intermediate state for a module. Please try it and tell me how it works on your side.

@TurboGit: I like working with #1419 applied. It makes undo much more useful for tonecurve/basecurve and gives me confidence to experiment a bit knowing that it is possible to pull back a change.

When I got carried away and made a mistake, I did still find myself skipping back and forth a bit to find the actual place to which I wanted to undo. For example, I made a good correction to one point on basecurve followed by a bad correction to another point. One thought would be to force recording of undo state whenever the user changes the point which they are manipulating in an iop, while also increasing the undo-record interval to eliminate recording so many in-between positions for each point. I'm not sure if the additional code complexity would be worth it, but it would model one idea of what an undo should do.

If it seems likely that #1419 or some variant will go in, it would make me happy as I would feel better about leaving the smooth scrolling behavior enabled for iops such as basecurve and tonecurve.

**Comment by LebedevRI** [comment]
Sorry, i keep meaning to re-review this pr, and then get carried away. Will try to look at it tomorrow.

**Comment by dtorop** [comment]
> Sorry, i keep meaning to re-review this pr, and then get carried away. Will try to look at it tomorrow.

Thank you, would appreciate the re-review as there is time!

**Comment by LebedevRI** [comment]
@dtorop see https://github.com/darktable-org/darktable/pull/1463

**Comment by LebedevRI** [comment]
@dtorop see https://redmine.darktable.org/issues/11594

**Pull request merged**